tests/jvm/src/test/scala/munit/AsyncFixtureOrderSuite.scala

```diff
@@ -0,0 +1,38 @@
+package munit
+
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class AsyncFixtureOrderSuite extends FunSuite {
+  val latch = Promise[Unit]
+  var completedFromTest = Option.empty[Boolean]
+  var completedFromTeardown = Option.empty[Boolean]
+
+  val latchOnTeardown = FunFixture.async[String](
+    setup = { test => Future.successful(test.name) },
+    teardown = { name =>
+      implicit val ec = munitExecutionContext
+      Future {
+        completedFromTeardown = Some(latch.trySuccess(()));
+      }
+    }
+  )
+
+  override def afterAll(): Unit = {
+    // promise was completed first form the test
+    assertEquals(completedFromTest, Some(true))
+    // and then from the teardown
+    assertEquals(completedFromTeardown, Some(false))
+  }
+
+  latchOnTeardown.test("teardown runs only after test completes") { _ =>
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future {
```

Comment:
With a guard comment that the test much be such that it eventually succeeds or fails, this becomes a test for latch.ready.

 If the latch is ready, the answer is in its Boolean value
and the Future is completed successfully with that value (synchronous).

If the latch is  not ready, the return a Future which contains a latch.foreach(mumble) or 
latch.map().  That Future should not complete until the latch is ready.  

A paranoid person, such as myself, would not trust the pre-condition of the Promise eventually completing and would add a timeout. Unfortunately, that adds another whole level of complexity.  In my world, test code should be robust to the situation were the whole purpose of the test is to try to break the world of execution.